crypto/pkcs11key/key_test.go

```diff
@@ -46,85 +99,219 @@ func (c mockCtx) GetAttributeValue(sh pkcs11.SessionHandle, o pkcs11.ObjectHandl
 	return sessionHandle, nil
 }
 
-func (c mockCtx) SignInit(sh pkcs11.SessionHandle, m []*pkcs11.Mechanism, o pkcs11.ObjectHandle) error {
+func (c *mockCtx) SignInit(sh pkcs11.SessionHandle, m []*pkcs11.Mechanism, o pkcs11.ObjectHandle) error {
 	return nil
 }
 
-func (c mockCtx) Sign(sh pkcs11.SessionHandle, message []byte) ([]byte, error) {
-	return []byte("some signed data"), nil
-}
-
-func TestSetup(t *testing.T) {
-	ps := &Key{
-		module:     mockCtx{},
-		tokenLabel: "token label",
-		pin:        "unused",
-	}
-	err := ps.setup("private key label")
-	if err != nil {
-		t.Errorf("Failed to set up Key: %s", err)
-	}
+func (c *mockCtx) Sign(sh pkcs11.SessionHandle, message []byte) ([]byte, error) {
+	return message, nil
 }
 
-func setup(t *testing.T) *Key {
+func setup(t *testing.T, label string) *Key {
 	ps := Key{
-		module:     mockCtx{},
+		module:     &mockCtx{},
 		tokenLabel: "token label",
 		pin:        "unused",
 	}
-	err := ps.setup("private key label")
+	err := ps.setup(label)
 	if err != nil {
 		t.Fatalf("Failed to set up Key: %s", err)
 	}
 	return &ps
 }
 
-func TestSign(t *testing.T) {
-	ps := setup(t)
+var signInput = []byte("1234567890 1234567890 1234567890")
+
+func sign(t *testing.T, ps *Key) []byte {
 	// Sign input must be exactly 32 bytes to match SHA256 size. In normally
 	// usage, Sign would be called by e.g. x509.CreateCertificate, which would
 	// handle padding to the necessary size.
-	signInput := []byte("1234567890 1234567890 1234567890")
 	output, err := ps.Sign(rand.Reader, signInput, crypto.SHA256)
 	if err != nil {
 		t.Fatalf("Failed to sign: %s", err)
 	}
-	if !bytes.Equal(output, []byte("some signed data")) {
+
+	if len(output) < len(signInput) {
+		t.Fatalf("Invalid signature size")
+	}
+
+	i := len(output) - len(signInput)
+	if !bytes.Equal(output[i:], signInput) {
 		t.Fatal("Incorrect sign output")
 	}
+	return output
+}
+
+func TestSign(t *testing.T) {
+	ps := setup(t, "rsa")
+	sig := sign(t, ps)
+
+	// Check that the RSA signature starts with the SHA256 hash prefix
+	var sha256Pre = []byte{0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86,
+		0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20}
+	if !(bytes.Equal(sha256Pre, sig[0:19])) {
+		t.Fatal("RSA signature doesn't start with prefix")
+	}
+
+	pub := ps.Public()
+	// Check public key is of right type
+	_, ok := pub.(*rsa.PublicKey)
+	if !ok {
+		t.Errorf("Attempted to load RSA key from module, got key of type %s. Expected *rsa.PublicKey", reflect.TypeOf(pub))
+	}
+
+	ps = setup(t, "ec")
+	sig = sign(t, ps)
+
+	if !(bytes.Equal(signInput, sig)) {
+		t.Fatal("ECDSA signature error")
+	}
+
+	pub = ps.Public()
+	// Check public key is of right type
+	ecPub, ok := pub.(*ecdsa.PublicKey)
+	if !ok {
+		t.Errorf("Attempted to load ECDSA key from module, got key of type %s. Expected *ecdsa.PublicKey", reflect.TypeOf(pub))
+	}
+	if ecPub == nil ||
+		!(bytes.Equal(ecPub.X.Bytes(), ecPoint[3:35]) &&
```

Comment:
The EC point in test has the ASN.1 wrapper already so that it tests the most code. I'm adding ASN.1 to the invalid key as well. I added some `readECPoint` tests.